include/lsst/afw/typehandling/GenericMap.h

```diff
@@ -495,32 +495,119 @@ class GenericMap {
+     *      * `std::string const&`
+     *      * `Storable const&`
+     *      * `std::shared_ptr<Storable>`
+     *
+     * @note This implementation calls @ref keys, then calls @ref unsafeLookup
+     *       for each key before passing the result to `visitor`.
+     *
+     * An example visitor that prints each key-value pair to standard output:
+     *
+     *     template <typename K>
+     *     class Printer {
+     *     public:
+     *         template <typename V>
+     *         void operator()(K const& key, V const& value) {
+     *             std::cout << key << ": " << value << "," << std::endl;
+     *         }
+     *
+     *         void operator()(K const& key, std::shared_ptr<Storable> value) {
+     *             std::cout << key << ": ";
+     *             if (value != nullptr)
+     *                 std::cout << *value;
+     *             else
+     *                 std::cout << "null";
+     *             std::cout << "," << std::endl;
+     *         }
+     *     };
+     */
+    template <class Visitor>
+    auto apply(Visitor&& visitor) const {
+        // Delegate to private methods to hide special-casing of Visitor
+        return _apply(visitor);
+    }
+
+    /**
+     * Apply a modifying operation to each key-value pair in the map.
+     *
+     * @tparam Visitor a callable that takes a key and a value. Requirements as for
+     *                 @ref apply(Visitor&&) const, except that it may take
+     *                 non-`const` references to values.
+     * @param visitor the visitor to apply
+     * @returns if `Visitor` has a return value, a `std::vector` of values
+     *          returned from applying `visitor` to each key in @ref keys, in
+     *          that order. Otherwise, `void`.
+     *
+     * @exceptsafe Provides basic exception safety if `Visitor` is exception-safe.
+     *
+     * @note This implementation calls @ref keys, then calls @ref unsafeLookup
+     *       for each key before passing the result to `visitor`.
+     */
+    template <class Visitor>
+    auto apply(Visitor&& visitor) {
+        // Delegate to private methods to hide special-casing of Visitor
+        return _apply(visitor);
+    }
+
 private:
     // Icky TMP, but I can't find another way to get at the template arguments for variant :(
-    // Method has no definition but can't be deleted without breaking definition of StorableType
+    // Methods have no definition but can't be deleted without breaking definition of StorableType
     /// @cond
+    template <typename T>
+    using _RemoveConstFromRef = std::add_lvalue_reference_t<std::remove_const_t<std::remove_reference_t<T>>>;
     template <typename... Types>
     static boost::variant<std::decay_t<Types>...> _referenceToType(boost::variant<Types...> const&) noexcept;
+    template <typename... Types>
+    static boost::variant<_RemoveConstFromRef<Types>...> _constRefToRef(
+            boost::variant<Types...> const&) noexcept;
```

Comment:
I want to keep the existing comment, because I *really* owe an apology to anyone with the misfortune of reading this code. 😭 

How about:

    // Transformations between value/reference/ref-to-const versions of internal variant type, to let
    //     the set of value types supported by GenericMap be defined only once
    // Icky TMP, but I can't find another way to get at the template arguments for variant :(
    // Methods have no definition but can't be deleted without breaking type definitions below

Reply:
Yes, that's helpful.